software/py/bsg_manycore_link_gen.py

```diff
@@ -0,0 +1,252 @@
+#!/usr/bin/python3
+
+import argparse, re
+
+"""
+BSG Manycore linker command file generator.
+
+Usage: ./bsg_manycore_link_gen.py -h
+
+Bandhav Veluri
+02/11/2020
+"""
+
+class bsg_manycore_link_gen:
+  """
+  A configurable linker command file generator for linking BSG Manycore
+  kernels. Kernel memory has private and shared regions backed by 4KB
+  on-tile sram (DMEM) and off-chip DRAM, respectively. Configuration options
+  include default location of data (dmem or dram), stack pointer and
+  dram memory size. These options won't affect the following:
+
+  1. Data attributed to .dmem* and .dram* sections. Following decalaration
+
+     int foo __attribute__((section (".dram")));
+
+     would place foo in dram region irrespective of default data location.
+
+  2. Data in bsg_manycore_lib would always be placed in dmem region.
+
+  
+  Note on program memory:
+  
+  Manycore's PC width is 24-bit, addressing a total of 16MB program space.
+  Since the 31:25 MBSs are assumed to be 0s, linking should also assume
+  0x0-0x01000000 as the valid program space. This means that .text section VMAs
+  lie in 0x0-0x01000000. On the other hand, .text section is loaded to DRAM and
+  as result, it's LMAs shouls be >0x80000000. So, .text section is loaded in the
+  region 0x80000000-0x81000000. Consequently, VMA's 0x80000000-0x81000000 are
+  are reserved for .text section as well.
+  """
+
+  _opening_comment = \
+    "/**********************************\n" \
+    + " BSG Manycore Linker Script \n\n"
+
+  def __init__(self, default_data_loc, dram_size, sp):
+    self._default_data_loc = default_data_loc
+    self._dram_size = dram_size
+    self._sp = sp
+    self._opening_comment += \
+        " data default: {0}\n".format(default_data_loc) \
+        + " dram memory size: 0x{0:08x}\n".format(dram_size) \
```

Comment:
Is there more information that we can put in this opening comment?

Reply:
For example, are all the options captured here? Like we do when running simulation, it might be useful to put as much information here as possible so that we can reproduce bugs/errors in a tool

Reply:
Missed stack pointer initialization... added it. That should cover all three configuration options. Also added time stamp.

Reply:
Cool, thanks!